File: kotlin/src/main/kotlin/chromahash/BitPack.kt
package chromahash

/** Write [count] bits of [value] starting at [bitpos] in little-endian byte order. */
internal fun writeBits(
    hash: ByteArray,
    bitpos: Int,
    count: Int,
    value: Int,
) {
    for (i in 0 until count) {
        val byteIdx = (bitpos + i) / 8
        val bitIdx = (bitpos + i) % 8
        if ((value ushr i) and 1 != 0) {
            hash[byteIdx] = (hash[byteIdx].toInt() or (1 shl bitIdx)).toByte()
        }
    }
}

/** Read [count] bits starting at [bitpos] in little-endian byte order. */
internal fun readBits(
    hash: ByteArray,
    bitpos: Int,
    count: Int,
): Int {
    var value = 0
    for (i in 0 until count) {
        val byteIdx = (bitpos + i) / 8
        val bitIdx = (bitpos + i) % 8
        if ((hash[byteIdx].toInt() and 0xFF) and (1 shl bitIdx) != 0) {
            value = value or (1 shl i)
        }
    }
    return value
}

File: csharp/src/Chromahash/BitPack.cs
namespace ChromaHash;

internal static class BitPack
{
    /// <summary>Write count bits of value starting at bitpos in little-endian byte order. Per spec §12.7 writeBits.</summary>
    public static void WriteBits(byte[] hash, int bitpos, int count, uint value)
    {
        for (int i = 0; i < count; i++)
        {
            int byteIdx = (bitpos + i) / 8;
            int bitIdx = (bitpos + i) % 8;
            if (((value >> i) & 1) != 0)
                hash[byteIdx] |= (byte)(1 << bitIdx);
        }
    }

    /// <summary>Read count bits starting at bitpos in little-endian byte order. Per spec §12.7 readBits.</summary>
    public static uint ReadBits(byte[] hash, int bitpos, int count)
    {
        uint value = 0;
        for (int i = 0; i < count; i++)
        {
            int byteIdx = (bitpos + i) / 8;
            int bitIdx = (bitpos + i) % 8;
            if ((hash[byteIdx] & (1 << bitIdx)) != 0)
                value |= (1u << i);
        }
        return value;
    }
}
